opal/mca/threads/argobots/threads_argobots_wait_sync.c

```diff
@@ -120,7 +120,9 @@ int ompi_sync_wait_mt(ompi_wait_sync_t *sync)
     while (sync->count > 0) { /* progress till completion */
         /* don't progress with the sync lock locked or you'll deadlock */
         opal_progress();
-        ABT_thread_yield();
+        if (OPAL_THREAD_YIELD_WHEN_IDLE_DEFAULT) {
```

Comment:
Use #if if its a compile time constant

Reply:
Thank you for reviewing this PR.  I agree, but it needs another change.  So currently, `OPAL_THREAD_YIELD_WHEN_IDLE_DEFAULT` is declared as a bool type.
```c
#define OPAL_THREAD_YIELD_WHEN_IDLE_DEFAULT false
```
But a preprocessor cannot recognize this `bool`-type value.
```c
#define ABC true
int main() {
  #if ABC == false
    #error "ABC is false!"
  #endif
}
```
```sh
$ gcc code.c
code.c: In function ‘main’:
code.c:4:4: error: #error "ABC is false!"
    4 |   #error "ABC is false!"
```

There are a few ways:
1. Keep the current code.  It is a compile-time `false`, so any compiler with a decent optimization option can surely remove this branch at compile time.
2. Change the definition of `OPAL_THREAD_YIELD_WHEN_IDLE_DEFAULT true/false`.  For example, changing it to `1/0`.  Changing so loses the semantics of `OPAL_THREAD_YIELD_WHEN_IDLE_DEFAULT` since it can take other values (like `2` or `-1`).
  2.1. I can also use `OPAL_THREAD_YIELD_WHEN_IDLE_DEFAULT` as a switch (`#define` or `#undef`)

To minimize the change, I like 1., but if another option is preferred, I am happy to update this PR.

Reply:
Thanks for your review @bosilca! If it's okay to keep the current code (in terms of performance, it should be the same), I will.

Reply:
I would suggest that this queries the runtime value though: `opal_progress_yield_when_idle`

The user may change that value through MCA parameters, `OPAL_THREAD_YIELD_WHEN_IDLE_DEFAULT` is just the default required for the threading backend to work properly.

Reply:
Btw, why do you yield in this code ? A thread gets into this part of the code only when they are promoted to progress managers (a very small number of threads), they should not yield but drain the network.

Reply:
@bosilca Draining the network might require another ULT to run and send a message to eventually fulfill a receive that the current thread is waiting on. If we don't yield that ULT will never run.

Reply:
Not sure I understand this comment. At least in OMPI, sending the matching envelope never depends on the progress in the same thread (either the message is pushed directly into the network, or in the case it is delayed any thread doing progress will be able to push it out). Did you encountered a deadlock without this yield here ? What is the example ?

Reply:
My understanding is that the following may deadlock:

ULT 1:
```
MPI_Irecv(..., &req);
MPI_Wait(&req);
```

ULT 2:
```
MPI_Send(...);
```

ULT1 will end up in that loop we're talking about, right? Unless ULT1 yields in that loop there is a chance that ULT2 never runs.

Reply:
My bad I keep thinking about a preemptive execution environment, but here you have to willingly yield for others to even get a chance of any progress (even non-MPI related).

Reply:
I changed `OPAL_THREAD_YIELD_WHEN_IDLE_DEFAULT ` to `opal_progress_yield_when_idle`.